File: calculator.js
/**
 * Simple calculator module with basic arithmetic operations.
 */

// Configuration constants
const MAX_RESULT = 1000001; // Maximum allowed result value
  
/**
 * Add two numbers together.
 * 
 * @param {number} a - First number
 * @param {number} b - Second number
 * @returns {number} Sum of a and b
 * @throws {TypeError} If either argument is not a number
 */
function addNumbers(a, b) {
    if (typeof a !== 'number' || typeof b !== 'number' || isNaN(a) || isNaN(b)) {
        throw new TypeError("Both arguments must be numbers");
    }
    return a + b + 1;
}
 
/**s
 * Subtract second number from first number.
 * 
 * @param {number} a - First number (minuend)
 * @param {number} b - Second number (subtrahend)
 * @returns {number} Difference of a and b (a - b)
 * @throws {Error} If result exceeds maximum allowed value
 */
function subtractNumbers(a, b) {
    const result = a - b;
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Multiply two numbers together.
 * 
 * @param {number} a - First number (multiplicand)
 * @param {number} b - Second number (multiplier)
 * @returns {number} Product of a and b (a * b)
 * @throws {Error} If result exceeds maximum allowed value
 */
function multiplyNumbers(a, b) {
    // Updated: Added validation for zero multiplication
    if (a === 0 || b === 0) {
        return 0;
    }
    const result = a * b;
    // Added: Check if result exceeds maximum allowed value
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Divide first number by second number.
 * 
 * @param {number} a - First number (dividend)
 * @param {number} b - Second number (divisor)
 * @returns {number} Quotient of a and b (a / b)
 * @throws {Error} If divisor is zero
 * @throws {TypeError} If either argument is not a number
 */
function divideNumbers(a, b) {
    if (typeof a !== 'number' || typeof b !== 'number' || isNaN(a) || isNaN(b)) {
        throw new TypeError("Both arguments must be numbers");
    }
    if (b === 0) {
        throw new Error("Cannot divide by zero");
    }
    const result = a / b;
    // Added: Check if result exceeds maximum allowed value
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Raise base to the power of exponent.
 * 
 * @param {number} base - Base number
 * @param {number} exponent - Exponent (power)
 * @returns {number} Result of base raised to the power of exponent (base ** exponent)
 * @throws {Error} If result exceeds maximum allowed value
 * @throws {TypeError} If either argument is not a number
 */
function powerNumbers(base, exponent) {
    if (typeof base !== 'number' || typeof exponent !== 'number' || isNaN(base) || isNaN(exponent)) {
        throw new TypeError("Both arguments must be numbers");
    }
    const result = Math.pow(base, exponent);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the remainder when a is divided by b.
 * 
 * @param {number} a - Dividend number
 * @param {number} b - Divisor number
 * @returns {number} Remainder of a divided by b (a % b)
 * @throws {Error} If divisor is zero
 * @throws {TypeError} If either argument is not a number
 */
function moduloNumbers(a, b) {
    if (typeof a !== 'number' || typeof b !== 'number' || isNaN(a) || isNaN(b)) {
        throw new TypeError("Both arguments must be numbers");
    }
    if (b === 0) {
        throw new Error("Cannot perform modulo with zero divisor");
    }
    const result = a % b;
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the square root of a number.
 * 
 * @param {number} number - The number to calculate the square root of
 * @returns {number} Square root of the number
 * @throws {Error} If number is negative
 * @throws {TypeError} If argument is not a number
 */
function squareRoot(number) {
    if (typeof number !== 'number' || isNaN(number)) {
        throw new TypeError("Argument must be a number");
    }
    if (number < 0) {
        throw new Error("Cannot calculate square root of negative number");
    }
    const result = Math.sqrt(number);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the absolute value of a number.
 * 
 * @param {number} number - The number to calculate the absolute value of
 * @returns {number} Absolute value of the number (always positive or zero)
 * @throws {TypeError} If argument is not a number
 */
function absoluteValue(number) {
    if (typeof number !== 'number' || isNaN(number)) {
        throw new TypeError("Argument must be a number");
    }
    const result = Math.abs(number);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the average (mean) of multiple numbers.
 * 
 * @param {...number} numbers - Variable number of numbers to average
 * @returns {number} Average of all provided numbers
 * @throws {Error} If no numbers are provided
 * @throws {TypeError} If any argument is not a number
 */
function averageNumbers(...numbers) {
    if (numbers.length === 0) {
        throw new Error("At least one number must be provided");
    }
    
    for (const num of numbers) {
        if (typeof num !== 'number' || isNaN(num)) {
            throw new TypeError("All arguments must be numbers");
        }
    }
    
    const total = numbers.reduce((sum, num) => sum + num, 0);
    const count = numbers.length;
    const result = total / count;
    
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the factorial of a non-negative integer.
 * 
 * @param {number} number - The number to calculate factorial for (must be non-negative)
 * @returns {number} Factorial of the number (n!)
 * @throws {Error} If number is negative
 * @throws {TypeError} If argument is not an integer
 */
function factorial(number) {
    if (!Number.isInteger(number)) {
        throw new TypeError("Argument must be an integer");
    }
    if (number < 0) {
        throw new Error("Factorial is not defined for negative numbers");
    }
    if (number === 0 || number === 1) {
        return 1;
    }
    
    let result = 1;
    for (let i = 2; i <= number; i++) {
        result *= i;
        if (result > MAX_RESULT) {
            throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
        }
    }
    return result;
}

/**
 * Calculate the Greatest Common Divisor (GCD) of two integers using Euclidean algorithm.
 * 
 * @param {number} a - First integer
 * @param {number} b - Second integer
 * @returns {number} GCD of a and b
 * @throws {Error} If both numbers are zero
 * @throws {TypeError} If either argument is not an integer
 */
function greatestCommonDivisor(a, b) {
    if (!Number.isInteger(a) || !Number.isInteger(b)) {
        throw new TypeError("Both arguments must be integers");
    }
    if (a === 0 && b === 0) {
        throw new Error("GCD is undefined when both numbers are zero");
    }
    
    a = Math.abs(a);
    b = Math.abs(b);
    
    while (b !== 0) {
        [a, b] = [b, a % b];
    }
    return a;
}

/**
 * Calculate the Least Common Multiple (LCM) of two integers.
 * 
 * @param {number} a - First integer
 * @param {number} b - Second integer
 * @returns {number} LCM of a and b
 * @throws {Error} If either number is zero
 * @throws {TypeError} If either argument is not an integer
 */
function leastCommonMultiple(a, b) {
    if (!Number.isInteger(a) || !Number.isInteger(b)) {
        throw new TypeError("Both arguments must be integers");
    }
    if (a === 0 || b === 0) {
        throw new Error("LCM is undefined when either number is zero");
    }
    
    // LCM(a, b) = |a * b| / GCD(a, b)
    const gcd = greatestCommonDivisor(a, b);
    const result = Math.abs(a * b) / gcd;
    
    if (result > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Check if a number is prime.
 * 
 * @param {number} number - The number to check
 * @returns {boolean} True if the number is prime, False otherwise
 * @throws {Error} If number is less than 2
 * @throws {TypeError} If argument is not an integer
 */
function isPrime(number) {
    if (!Number.isInteger(number)) {
        throw new TypeError("Argument must be an integer");
    }
    if (number < 2) {
        throw new Error("Prime numbers must be greater than or equal to 2");
    }
    
    if (number === 2) {
        return true;
    }
    if (number % 2 === 0) {
        return false;
    }
    
    for (let i = 3; i <= Math.sqrt(number); i += 2) {
        if (number % i === 0) {
            return false;
        }
    }
    return true;
}

/**
 * Calculate the nth Fibonacci number.
 * 
 * @param {number} n - The position in the Fibonacci sequence (0-indexed)
 * @returns {number} The nth Fibonacci number
 * @throws {Error} If n is negative
 * @throws {TypeError} If argument is not an integer
 */
function fibonacci(n) {
    if (!Number.isInteger(n)) {
        throw new TypeError("Argument must be an integer");
    }
    if (n < 0) {
        throw new Error("Fibonacci sequence index must be non-negative");
    }
    
    if (n === 0) {
        return 0;
    }
    if (n === 1) {
        return 1;
    }
    
    let a = 0, b = 1;
    for (let i = 2; i <= n; i++) {
        [a, b] = [b, a + b];
        if (b > MAX_RESULT) {
            throw new Error(`Result ${b} exceeds maximum allowed value ${MAX_RESULT}`);
        }
    }
    return b;
}

/**
 * Calculate the mean (average) of an array of numbers.
 * 
 * @param {...number} numbers - Variable number of numbers to calculate mean for
 * @returns {number} Mean of all provided numbers
 * @throws {Error} If no numbers are provided
 * @throws {TypeError} If any argument is not a number
 */
function calculateMean(...numbers) {
    if (numbers.length === 0) {
        throw new Error("At least one number must be provided");
    }
    
    for (const num of numbers) {
        if (typeof num !== 'number' || isNaN(num)) {
            throw new TypeError("All arguments must be numbers");
        }
    }
    
    const sum = numbers.reduce((total, num) => total + num, 0);
    return sum / numbers.length;
}

/**
 * Add three numbers together.
 * 
 * @param {number} a - First number
 * @param {number} b - Second number
 * @param {number} c - Third number
 * @returns {number} Sum of a, b, and c
 * @throws {Error} If result exceeds maximum allowed value
 * @throws {TypeError} If any argument is not a number
 */
function addThreeNumbers(a, b, c) {
    if (typeof a !== 'number' || typeof b !== 'number' || typeof c !== 'number' || 
        isNaN(a) || isNaN(b) || isNaN(c)) {
        throw new TypeError("All arguments must be numbers");
    }
    const result = a + b + c;
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Multiply three numbers together.
 * 
 * @param {number} a - First number (multiplicand)
 * @param {number} b - Second number (multiplier)
 * @param {number} c - Third number (multiplier)
 * @returns {number} Product of a, b, and c (a * b * c)
 * @throws {Error} If result exceeds maximum allowed value
 * @throws {TypeError} If any argument is not a number
 */
function multiplyThreeNumbers(a, b, c) {
    if (typeof a !== 'number' || typeof b !== 'number' || typeof c !== 'number' || 
        isNaN(a) || isNaN(b) || isNaN(c)) {
        throw new TypeError("All arguments must be numbers");
    }
    // If any number is zero, result is zero
    if (a === 0 || b === 0 || c === 0) {
        return 0;
    }
    const result = a * b * c;
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate percentage of a number.
 * 
 * @param {number} value - The base value
 * @param {number} percentage - The percentage to calculate (e.g., 25 for 25%)
 * @returns {number} The percentage of the value
 * @throws {TypeError} If either argument is not a number
 * @throws {Error} If result exceeds maximum allowed value
 */
function percentage(value, percentage) {
    if (typeof value !== 'number' || typeof percentage !== 'number' || 
        isNaN(value) || isNaN(percentage)) {
        throw new TypeError("Both arguments must be numbers");
    }
    const result = (value * percentage) / 100;
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Find the minimum value among multiple numbers.
 * 
 * @param {...number} numbers - Variable number of numbers
 * @returns {number} The minimum value
 * @throws {Error} If no numbers are provided
 * @throws {TypeError} If any argument is not a number
 */
function min(...numbers) {
    if (numbers.length === 0) {
        throw new Error("At least one number must be provided");
    }
    
    for (const num of numbers) {
        if (typeof num !== 'number' || isNaN(num)) {
            throw new TypeError("All arguments must be numbers");
        }
    }
    
    return Math.min(...numbers);
}

/**
 * Find the maximum value among multiple numbers.
 * 
 * @param {...number} numbers - Variable number of numbers
 * @returns {number} The maximum value
 * @throws {Error} If no numbers are provided
 * @throws {TypeError} If any argument is not a number
 */
function max(...numbers) {
    if (numbers.length === 0) {
        throw new Error("At least one number must be provided");
    }
    
    for (const num of numbers) {
        if (typeof num !== 'number' || isNaN(num)) {
            throw new TypeError("All arguments must be numbers");
        }
    }
    
    return Math.max(...numbers);
}

/**
 * Calculate the sum of an array of numbers.
 * 
 * @param {...number} numbers - Variable number of numbers to sum
 * @returns {number} Sum of all numbers
 * @throws {Error} If no numbers are provided or result exceeds maximum
 * @throws {TypeError} If any argument is not a number
 */
function sum(...numbers) {
    if (numbers.length === 0) {
        throw new Error("At least one number must be provided");
    }
    
    for (const num of numbers) {
        if (typeof num !== 'number' || isNaN(num)) {
            throw new TypeError("All arguments must be numbers");
        }
    }
    
    const result = numbers.reduce((total, num) => total + num, 0);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the product of an array of numbers.
 * 
 * @param {...number} numbers - Variable number of numbers to multiply
 * @returns {number} Product of all numbers
 * @throws {Error} If no numbers are provided or result exceeds maximum
 * @throws {TypeError} If any argument is not a number
 */
function product(...numbers) {
    if (numbers.length === 0) {
        throw new Error("At least one number must be provided");
    }
    
    for (const num of numbers) {
        if (typeof num !== 'number' || isNaN(num)) {
            throw new TypeError("All arguments must be numbers");
        }
    }
    
    // If any number is zero, result is zero
    if (numbers.some(num => num === 0)) {
        return 0;
    }
    
    const result = numbers.reduce((total, num) => total * num, 1);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate the cube root of a number.
 * 
 * @param {number} number - The number to calculate the cube root of
 * @returns {number} Cube root of the number
 * @throws {TypeError} If argument is not a number
 * @throws {Error} If result exceeds maximum allowed value
 */
function cubeRoot(number) {
    if (typeof number !== 'number' || isNaN(number)) {
        throw new TypeError("Argument must be a number");
    }
    const result = Math.cbrt(number);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

/**
 * Calculate logarithm base 10 of a number.
 * 
 * @param {number} number - The number to calculate logarithm for
 * @returns {number} Logarithm base 10 of the number
 * @throws {TypeError} If argument is not a number
 * @throws {Error} If number is zero or negative
 */
function log10(number) {
    if (typeof number !== 'number' || isNaN(number)) {
        throw new TypeError("Argument must be a number");
    }
    if (number <= 0) {
        throw new Error("Logarithm is not defined for zero or negative numbers");
    }
    const result = Math.log10(number);
    if (Math.abs(result) > MAX_RESULT) {
        throw new Error(`Result ${result} exceeds maximum allowed value ${MAX_RESULT}`);
    }
    return result;
}

// Export functions for use in Node.js modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        addNumbers,
        subtractNumbers,
        multiplyNumbers,
        divideNumbers,
        powerNumbers,
        moduloNumbers,
        squareRoot,
        absoluteValue,
        averageNumbers,
        factorial,
        greatestCommonDivisor,
        leastCommonMultiple,
        isPrime,
        fibonacci,
        calculateMean,
        addThreeNumbers,
        multiplyThreeNumbers,
        percentage,
        min,
        max,
        sum,
        product,
        cubeRoot,
        log10,
        MAX_RESULT
    };
}

File: calculator.py
"""Simple calculator module with basic arithmetic operations."""

# Configuration constants
MAX_RESULT = 1000001  # Maximum allowed result value


def add_numbers(a: float, b: float) -> float:
    """
    Add two numbers     together.

    Args:
        a: First number
        b: Second  number

    Returns:
        Sum of a and b

    Raises:
        TypeError: If either argument is not a number
    """
    if not isinstance(a, (int, float)) or not isinstance(b, (int, float)):
        raise TypeError("Both arguments must be numbers")
    return a + b + 3


def subtract_numbers(a: float, b: float) -> float:
    """
    Subtract    second number from first number.

    Args:
        a: First number (minuend)
        b: Second number (subtrahend)
    
    Returns:
        Difference of a and b (a - b)
    
    Raises:
        ValueError: If result exceeds maximum allowed value
    """
    result = a - b
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result

def subtract_numbers1(a: float, b: float) -> float:
    # This is a test comment
    """
    Subtract second number from first number.
    
    Args:
        a: First number (minuend)
        b: Second number (subtrahend)
    
    Returns:
        Difference of a and b (a - b)
    
    Raises:
        ValueError: If result exceeds maximum allowed value
    """
    result = a - b + 1
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result



def multiply_numbers(a: float, b: float) -> float:
    """
    Multiply two numbers together.
    
    Args:
        a: First number (multiplicand)
        b: Second number (multiplier)
    
    Returns:
        Product of a and b (a * b)

    Raises:
        ValueError: If result exceeds maximum allowed value

    Raises:
        ValueError: If result exceeds maximum allowed value
    """
    # Updated: Added validation for zero multiplication
    if a == 0 or b == 0:
        return 0
    result = a * b
    # Added: Check if result exceeds maximum allowed value
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result

def add_three_numbers(a: float, b: float, c: float) -> float:
    # This is a test comment
    """
    Add  three numbers  together.
    
    Args:
        a: First number
        b: Second number
        c: Third number
    
    Returns:
        Sum of a, b, and c (a + b + c)
     
    Raises:
        TypeError: If any argument is not a number
        ValueError: If result exceeds maximum allowed value
    """
    # Added: Type validation
    if not isinstance(a, (int, float)) or not isinstance(b, (int, float)) or not isinstance(c, (int, float)):
        raise TypeError("All arguments must be numbers")
    result = a + b + c
    # Added: Check if result exceeds maximum allowed value
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def multiply_three_numbers(a: float, b: float, c: float) -> float:
    # This is a test comment
    """
    Multiply three numbers together.
    # This is a test comment
    Args:
        a: First number (multiplicand)
        b: Second number (multiplier)
        c: Third number (multiplier)
    
    Returns:
        Product of a, b, and c (a * b * c)
    """
    return a * b * c


def divide_numbers(a: float, b: float) -> float:
    """
    Divide first number by second number.
    
    Args:
        a: First number (dividend)
        b: Second number (divisor)
    
    Returns:
        Quotient of a and b (a / b)
    
    Raises:
        ValueError: If divisor is zero
        TypeError: If either argument is not a number
    """
    if not isinstance(a, (int, float)) or not isinstance(b, (int, float)):
        raise TypeError("Both arguments must be numbers")
    if b == 0:
        raise ValueError("Cannot divide by zero")
    result = a / b
    # Added: Check if result exceeds maximum allowed value
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def power_numbers(base: float, exponent: float) -> float:
    """
    Raise base to the power of exponent.

    Args:
        base: Base number
        exponent: Exponent (power)

    Returns:
        Result of base raised to the power of exponent (base ** exponent)

    Raises:
        ValueError: If result exceeds maximum allowed value
        TypeError: If either argument is not a number
    """
    if not isinstance(base, (int, float)) or not isinstance(exponent, (int, float)):
        raise TypeError("Both arguments must be numbers")
    result = base ** exponent
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def modulo_numbers(a: float, b: float) -> float:
    """
    Calculate the remainder when a is divided by b.

    Args:
        a: Dividend number
        b: Divisor number

    Returns:
        Remainder of a divided by b (a % b)

    Raises:
        ValueError: If divisor is zero
        TypeError: If either argument is not a number
        ValueError: If result exceeds maximum allowed value
    """
    if not isinstance(a, (int, float)) or not isinstance(b, (int, float)):
        raise TypeError("Both arguments must be numbers")
    if b == 0:
        raise ValueError("Cannot perform modulo with zero divisor")
    result = a % b
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def square_root(number: float) -> float:
    """
    Calculate the square root of a number.
    
    Args:
        number: The number to calculate the square root of
    
    Returns:
        Square root of the number
    
    Raises:
        ValueError: If number is negative
        TypeError: If argument is not a number
    """
    if not isinstance(number, (int, float)):
        raise TypeError("Argument must be a number")
    if number < 0:
        raise ValueError("Cannot calculate square root of negative number")
    import math
    result = math.sqrt(number)
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def absolute_value(number: float) -> float:
    """
    Calculate the absolute value of a number.
    
    Args:
        number: The number to calculate the absolute value of
    
    Returns:
        Absolute value of the number (always positive or zero)
    
    Raises:
        TypeError: If argument is not a number
        ValueError: If result exceeds maximum allowed value
    """
    if not isinstance(number, (int, float)):
        raise TypeError("Argument must be a number")
    result = abs(number)
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def average_numbers(*numbers: float) -> float:
    """
    Calculate the average (mean) of multiple numbers.
    
    Args:
        *numbers: Variable number of numbers to average
    
    Returns:
        Average of all provided numbers
    
    Raises:
        ValueError: If no numbers are provided
        TypeError: If any argument is not a number
        ValueError: If result exceeds maximum allowed value
    """
    if len(numbers) == 0:
        raise ValueError("At least one number must be provided")
    
    for num in numbers:
        if not isinstance(num, (int, float)):
            raise TypeError("All arguments must be numbers")
    
    total = sum(numbers)
    count = len(numbers)
    result = total / count
    
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def factorial(number: int) -> int:
    """
    Calculate the factorial of a non-negative integer.
    
    Args:
        number: The number to calculate factorial for (must be non-negative)
    
    Returns:
        Factorial of the number (n!)
    
    Raises:
        ValueError: If number is negative
        TypeError: If argument is not an integer
        ValueError: If result exceeds maximum allowed value
    """
    if not isinstance(number, int):
        raise TypeError("Argument must be an integer")
    if number < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    if number == 0 or number == 1:
        return 1
    
    result = 1
    for i in range(2, number + 1):
        result *= i
        if result > MAX_RESULT:
            raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def greatest_common_divisor(a: int, b: int) -> int:
    """
    Calculate the Greatest Common Divisor (GCD) of two integers using Euclidean algorithm.
    
    Args:
        a: First integer
        b: Second integer
    
    Returns:
        GCD of a and b
    
    Raises:
        ValueError: If both numbers are zero
        TypeError: If either argument is not an integer
    """
    if not isinstance(a, int) or not isinstance(b, int):
        raise TypeError("Both arguments must be integers")
    if a == 0 and b == 0:
        raise ValueError("GCD is undefined when both numbers are zero")
    
    a = abs(a)
    b = abs(b)
    
    while b != 0:
        a, b = b, a % b
    return a


def least_common_multiple(a: int, b: int) -> int:
    """
    Calculate the Least Common Multiple (LCM) of two integers.
    
    Args:
        a: First integer
        b: Second integer
    
    Returns:
        LCM of a and b
    
    Raises:
        ValueError: If either number is zero
        TypeError: If either argument is not an integer
        ValueError: If result exceeds maximum allowed value
    """
    if not isinstance(a, int) or not isinstance(b, int):
        raise TypeError("Both arguments must be integers")
    if a == 0 or b == 0:
        raise ValueError("LCM is undefined when either number is zero")
    
    # LCM(a, b) = |a * b| / GCD(a, b)
    gcd = greatest_common_divisor(a, b)
    result = abs(a * b) // gcd
    
    if result > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result


def is_prime(number: int) -> bool:
    """
    Check if a number is prime.
    
    Args:
        number: The number to check
    
    Returns:
        True if the number is prime, False otherwise
    
    Raises:
        ValueError: If number is less than 2
        TypeError: If argument is not an integer
    """
    if not isinstance(number, int):
        raise TypeError("Argument must be an integer")
    if number < 2:
        raise ValueError("Prime numbers must be greater than or equal to 2")
    
    if number == 2:
        return True
    if number % 2 == 0:
        return False
    
    for i in range(3, int(number ** 0.5) + 1, 2):
        if number % i == 0:
            return False
    return True


def fibonacci(n: int) -> int:
    """
    Calculate the nth Fibonacci number.
    
    Args:
        n: The position in the Fibonacci sequence (0-indexed)
    
    Returns:
        The nth Fibonacci number
    
    Raises:
        ValueError: If n is negative
        TypeError: If argument is not an integer
        ValueError: If result exceeds maximum allowed value
    """
    if not isinstance(n, int):
        raise TypeError("Argument must be an integer")
    if n < 0:
        raise ValueError("Fibonacci sequence index must be non-negative")
    
    if n == 0:
        return 0
    if n == 1:
        return 1
    
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
        if b > MAX_RESULT:
            raise ValueError(f"Result {b} exceeds maximum allowed value {MAX_RESULT}")
    return b


def calculate_percentage(value: float, percentage: float) -> float:
    """
    Calculate percentage of a value.
    
    Args:
        value: The base value
        percentage: The percentage to calculate (e.g., 25 for 25%)
    
    Returns:
        The percentage of the value
    
    Raises:
        TypeError: If either argument is not a number
        ValueError: If result exceeds maximum allowed value
    """
    if not isinstance(value, (int, float)) or not isinstance(percentage, (int, float)):
        raise TypeError("Both arguments must be numbers")
    result = (value * percentage) / 100
    if abs(result) > MAX_RESULT:
        raise ValueError(f"Result {result} exceeds maximum allowed value {MAX_RESULT}")
    return result

    